Week8

Kremifany opened the pull request:
Preferred home work assignment
added basic UI to show that back end calls are legitimate

Kremifany committed:
jwt token created

Kremifany committed:
authorized added

Kremifany committed:
authentication middleware created

Kremifany committed:
Homework in progress

Kremifany committed:
preffered assignment

mattcarlotta commented on 05-JWT-Basics/preferred/errors/unauthenticated.js:
```diff
@@ -0,0 +1,12 @@
+const CustomAPIError = require("./custom-error")
+const { StatusCodes } = require("http-status-codes");
+
+const UnauthenticatedError = require("./custom-error");
```
Requires change:
You're requiring this (`"./custom-error"`) import twice. Also, you're also exporting the `UnauthenticatedError` class instead of the `BadRequest` class.

mattcarlotta commented on 05-JWT-Basics/preferred/middleware/auth.js:
```diff
@@ -0,0 +1,29 @@
+const {UnauthenticatedError} = require("../errors");
+
+const authanticationMiddleware = async (req, res, next) => {
+    const { authorization } = req.headers;
+    console.log("Authorization", authorization);
+    console.log(req.headers);
+    if (!authorization || !authorization.startsWith("Bearer ")) {
+      throw new UnauthenticatedError("No token provided");
+    }
+
+    const token = authorization.split(" ")[1];
+    console.log(token);
+    try {
+      const decoded = jwt.verify(token, process.env.JWT_SECRET);
+      const {name} = decoded;
+      console.log(`name is: ${name}`);
+      console.log(`inside auth middleware${name}`)
+      req.user = {name}
+      next();
+    } catch (error) {
+      throw new UnauthenticatedError(
+        "Not authorizad to access this route"
+      );
+    }
+  
```
Nitpick:
This token validation logic can be simplified with [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). That said, we don't really care if the header and token is there, instead we only really care if the header token is valid!
```js
// optional chaining allows you to access a property that may or may not be there
// if it's not there, it'll fall back to undefined
const token = req.headers.authorization?.split(" ")?.[1];

try {
  req.user = jwt.verify(token, process.env.JWT_SECRET);
  next();
} catch {
  throw new UnauthenticatedError("Not authorized");
}
```
---
Alternatively, you can skip using optional chaining and instead just try to access the property within the `try` block:
```js
try {
  // this will throw an error if the authorization header is undefined
  const token = req.headers.authorization.split(" ")[1];
  req.user = jwt.verify(token, process.env.JWT_SECRET);
  next();
} catch {
  throw new UnauthenticatedError("Not authorized");
}
```

mattcarlotta commented on 05-JWT-Basics/preferred/public/browser-app.js:
Nitpick:
Any time you call a `document` search method (like `querySelector`), you're not guaranteed to get a HTMLElement. Instead, it can either be a HTMLElement or undefined. A good practice is to assume it's undefined and you'll add some conditional logic to only add event listeners when they're defined.

mattcarlotta replied:
On a related note, storing the token is localStorage can be a security vulnerability. Also, you'll have to add additional logic to check if the token has expired. This is why I like storing tokens in cookies because the browser takes care of the expiration and depending on the cookie attributes can prevent tampering.

mattcarlotta reviewed (approved):
Excellent work, Fany! You nearly aced this assignment. This only one required change that I've found, but that change doesn't involve another round of review. Please fix this. Otherwise, I've left some nitpicky feedback that you can address or leave as is. 👍 

I've also only reviewed the "preferred" code since the "starter" code has the same duplicate logic & mistakes.

Kremifany committed:
-bug fix, wrong Error class name.
-token validation simplified with conditional chainong
-added condition for adding event listeners to HTMLElement
-token expiration check added

Kremifany committed:
same message

mattcarlotta reviewed (approved):
Looks good to me! 👍
